adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,928 @@
+
+_Effects (1)_: Equivalent to [code]#h.update_host(acc)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { update_host(h,
+acc); })#.
+
+'''
+
+.[apititle]#prefetch#
+[source,role=synopsis,id=api:prefetch]
+----
+namespace sycl::khr {
+
+void prefetch(sycl::handler& h, void* ptr, size_t numBytes); (1)
+
+void prefetch(sycl::queue q, void* ptr, size_t numBytes);    (2)
+
+}
+----
+_Effects (1)_: Equivalent to [code]#h.prefetch(ptr, numBytes)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { prefetch(h, ptr,
+numBytes); })#.
+
+'''
+
+.[apititle]#mem_advise#
+[source,role=synopsis,id=api:mem_advise]
+----
+namespace sycl::khr {
+
+void mem_advise(sycl::handler& h, void* ptr, size_t numBytes, int advice); (1)
+
+void mem_advise(sycl::queue q, void* ptr, size_t numBytes, int advice);    (2)
+
+}
+----
+_Effects (1)_: Equivalent to [code]#h.mem_advise(ptr, numBytes, advice)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { mem_advise(h,
+ptr, numBytes, advice); })#.
+
+'''
+
+=== Command and event barriers
+
+.[apititle]#command_barrier#
+[source,role=synopsis,id=api:command_barrier]
+----
+namespace sycl::khr {
+
+void command_barrier(sycl::handler& h); (1)
+
+void command_barrier(sycl::queue q);    (2)
+
+}
+----
+_Effects_: Enqueues a command barrier.
+Any commands submitted after this barrier cannot begin execution until all
+previously submitted commands (and any commands associated with dependendent
+events) have completed.
+
+'''
+
+.[apititle]#event_barrier#
+[source,role=synopsis,id=api:event_barrier]
+----
+namespace sycl::khr {
+
+void event_barrier(sycl::handler& h, const std::vector<sycl::event>& events); (1)
+
+void event_barrier(sycl::queue q, const std::vector<sycl::event>& events);    (2)
+
+}
+----
+_Effects_: Enqueues an event barrier.
+Any commands submitted after this barrier cannot begin execution until all
+commands associated with [code]#events# (and any commands associated with other
+dependent events) have completed.
+
+{note}For both overloads, if [code]#events# is empty and an event barrier has no
+other dependencies (e.g., specified by [code]#handler::depends_on#), it is ot
+required to wait for any commands unless the [code]#queue# is in-order.{endnote}
+
+'''
+
+== [code]#invocation# class template
+
+The [code]#invocation# class template identifies an invocation of a kernel
+function.
+
+Instances of the [code]#invocation# class template are not user-constructible
+and are passed as an argument to each invocation of a kernel function.
+
+[source,role=synopsis]
+----
+namespace sycl::khr {
+
+template <int Dimensions>
+class invocation
+{
+ public:
+  static constexpr int dimensions = Dimensions;
+
+  id<Dimensions> id() const noexcept;
+  size_t linear_id() const noexcept;
+
+  range<Dimensions> range() const noexcept;
+
+  group<Dimensions> get_work_group() const noexcept;
+
+  sub_group<Dimensions> get_sub_group() const noexcept;
```

Comment:
I don't think so.  And -- annoyingly -- ISO C++ doesn't seem to be consistent, so we can't really rely on precedent.

The convention I've tried to follow when sketching out these KHRs is to use `get_` as a syntactic hint that an implementation is either 1) doing more than just returning a member of a class; and/or 2) that the thing being returned conceptually exists outside of the class.

So `invocation::id()` and `invocation::range()` don't have the `get_` prefix because instantiating an `id` and a `range` within each `invocation` is likely to be a common implementation choice, and because these properties conceptually "belong to" [/handwave] each invocation.

`invocation::get_group()` and `invocation::get_sub_group()` have the `get_` prefix because implementations might have to look-up which group an invocation belongs to and construct a `group`/`sub_group` class, and because the groups themselves conceptually exist outside of the invocation.

Although we don't have a clear precedent to follow, this at least seemed closely aligned to how several ISO C++ classes behave.  For example, in `std::vector`, the pointer returned by `data()` "belongs to" the class, whereas the allocator returned by `get_allocator()` conceptually exists outside of the vector.

Reply:
> The convention I've tried to follow when sketching out these KHRs is to use get_ as a syntactic hint that an implementation is either 1) doing more than just returning a member of a class; and/or 2) that the thing being returned conceptually exists outside of the class.

I suspected something along those lines; I think it's a good way of handling it!